src/include/duckdb/storage/buffer/buffer_pool.hpp

```diff
@@ -85,10 +88,35 @@ class BufferPool {
 	void IncrementDeadNodes(FileBufferType type);
 
 protected:
+	struct MemoryUsageCounters {
+		//! The maximum difference between memory statistics and actual usage is 2MB (64 * 32k)
+		static constexpr size_t kCacheCnts = 64;
+		static constexpr size_t kCacheThreshold = 32 << 10;
+		using MemoryUsagePerTag = std::array<atomic<int64_t>, MEMORY_TAG_COUNT>;
+
+		//! global memory usage counters
+		atomic<int64_t> memory_usage;
+		MemoryUsagePerTag memory_usage_per_tag;
+		//! cache memory usage to improve performance
+		std::array<MemoryUsagePerTag, kCacheCnts> memory_usage_caches;
+
+		MemoryUsageCounters();
+
+		idx_t GetUsedMemory() const {
+			auto used_memory = memory_usage.load(std::memory_order_relaxed);
+			return used_memory > 0 ? static_cast<idx_t>(used_memory) : 0;
```

Comment:
I would perhaps make a distinction here between the usage of `GetUsedMemory()` for evicting blocks (which is done frequently) - and getting the used memory for display (e.g. as part of `PRAGMA database_size` or `duckdb_memory()`).

Perhaps we can introduce an additional `GetUsedMemoryFlush()` method that is used in the latter scenario - in which case we flush the caches prior to reading out the `used_memory`.

While that might still provide a slightly wrong memory usage in concurrent situations, it would provide the correct number if other threads are not currently actively allocating memory (i.e. "at rest" memory is always accurate). This would then also mean the tests don't need to be changed anymore to account for the new inaccuracies.

Reply:
A `flush` parameter has been added to `GetUsedMemory` to indicate whether cache counters need to be flushed. Except for evicting blocks, cache counters will be flushed before reading. The structure of cache has also been modified to make 
 flush operation more lightweight.